scipy/spatial/_hausdorff.pyx

```diff
@@ -16,56 +16,73 @@ import numpy as np
 cimport numpy as np
 cimport cython
 from libc.math cimport sqrt
+from libc.stdlib cimport srand, rand, RAND_MAX
+from libc.limits cimport UINT_MAX
 
 __all__ = ['directed_hausdorff']
 
+cdef extern from "numpy/npy_math.h":
+    double inf "NPY_INFINITY"
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+@cython.cdivision(True)
+cdef void index_shuffler(long [::1] arr, size_t size):
+    cdef:
+        size_t i = 0
+        size_t j
+        int t
+
+    while i < (size - 1):
+        j = i + rand() / (RAND_MAX / (size - i) + 1)
+        t = arr[j]
+        arr[j] = arr[i]
+        arr[i] = t
+        i += 1
+
 @cython.boundscheck(False)
-def directed_hausdorff(double[:,::1] ar1, double[:,::1] ar2, seed=0):
+def directed_hausdorff(const double[:,::1] ar1,
+                       const double[:,::1] ar2,
+                       seed=0):
 
     cdef double cmax, cmin, d
-    cdef bint no_break_occurred
     cdef int N1 = ar1.shape[0]
     cdef int N2 = ar2.shape[0]
     cdef int data_dims = ar1.shape[1]
     cdef unsigned int i, j, k
-    cdef unsigned int i_store = 0, j_store = 0, i_ret = 0, j_ret = 0
-    cdef long[:] resort1, resort2
+    cdef unsigned int i_ret = 0, j_ret = 0
+    cdef long[::1] resort1 = np.arange(N1)
+    cdef long[::1] resort2 = np.arange(N2)
 
     # shuffling the points in each array generally increases the likelihood of
     # an advantageous break in the inner search loop and never decreases the
     # performance of the algorithm
     rng = np.random.RandomState(seed)
-    resort1 = np.arange(N1)
-    resort2 = np.arange(N2)
-    rng.shuffle(resort1)
-    rng.shuffle(resort2)
-    ar1 = np.asarray(ar1)[resort1]
-    ar2 = np.asarray(ar2)[resort2]
+    srand(rng.randint(UINT_MAX, dtype=np.uint64))
```

Comment:
there was considerable discussion in the original PR for the implementation re: the effect on random seeds in workflows, so this should be carefully scrutinized; again, not really needed to speed up larger data sets, but used to reduce overhead for the small ones

Reply:
When randomgen gets in, it will be possible to use a basic RNG as the source of randomness, so that you can have something like 

```
rng = Xoroshiro128()
rng = <brng_t *> PyCapsule_GetPointer(rng.capsule, capsule_name)
rng.next_uint53(rng.state)
```

with minimal overhead.  AFAIK the low-level shuffler is not directly exposed.